### PR TITLE
FantLab.ru BookFinder

### PR DESCRIPTION
**Pull request by Nab0y**
I've added a new book match provider. [fantlab.ru](https://fantlab.ru/) is a huge site with tens of thousands of book descriptions in Russian and not only (Russian is still the main language).
This is my first js code, my main activity is a .net developer. So if there are simple errors in the implementation, then please do not kick hard. :)
image
image

### CODE REVIEW

File: server/providers/FantLab.js [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+    })
+
+    return Promise.all(items.map(async item => await this.getWork(item)))
+  }
+
+  async getWork(item) {
+    var { work_id, work_type_id } = item
+
+    if (this._filterWorkType.includes(work_type_id)) {
+      return { title: null }
```
Comment (advplyr) [inline server/providers/FantLab.js:48]: These need to get filtered out before returning the results in the `search`

File: server/providers/FantLab.js [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+
+    var url = `${this._baseUrl}/work/${work_id}/extended`
+    var bookData = await axios.get(url).then((resp) => {
+      return resp.data || null
+    }).catch((error) => {
+      Logger.error(`[FantLab] work info reques error`, error)
+      return null
+    })
+
+    return await this.cleanBookData(bookData)
```
Comment (advplyr) [inline server/providers/FantLab.js:59]:
```suggestion
    return this.cleanBookData(bookData)
```

File: server/providers/FantLab.js [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+
+    var genres = classificatory ? this.tryGetGenres(classificatory) : []
+
+    var imageAndIsbn = await this.tryGetCoverFromEditions(editions_blocks)
+
+    if (imageAndIsbn) {
+      var { imageUrl, isbn } = imageAndIsbn
+
+      if (imageUrl) {
+        image = imageUrl
```
Comment (advplyr) [inline server/providers/FantLab.js:78]: I'm confused here. `image` is defined above from `bookData` but gets redefined here.

File: server/providers/FantLab.js [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+      isbn: isbn
+    }
+  }
+
+  tryGetGenres(classificatory) {
+    var { genre_group } = classificatory;
+    if (!genre_group) {
+      return []
+    }
+    var genresGroup = genre_group.find(group => group.genre_group_id = 1) // genres and subgenres
```
Comment (advplyr) [inline server/providers/FantLab.js:103]:
```suggestion
    const genresGroup = genre_group.find(group => group.genre_group_id == 1) // genres and subgenres
```

File: server/providers/FantLab.js [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+    // genre_group_id=5 - Recommended age of the reader
+
+    if (!genresGroup) return []
+
+    var { genre } = genresGroup;
+    var rootGenre = genre[0];
+
+    var { label } = rootGenre
+
+    return [label].concat(this.tryGetSubGenres(rootGenre))
```
Comment (advplyr) [inline server/providers/FantLab.js:119]: Do we need to separate this out into a sub genres function?
Could we get them all like this `genre.map(genreObj => genreObj.label)`

**Comment by advplyr** [comment]
Thanks! This one is tough for me because the API docs are in Russian.

Can you clean up:
1. Replace `var` with `const` or `let`. I recently stopped using `var` and have been replacing this slowly throughout the project. `var` can be instantiated inside an `if` statement but accessible outside the `if` statement which can be confusing. See line 75 in your PR.
2. Removing semicolons

File: server/providers/FantLab.js (hunk 1 above)
Comment (Nab0y) [reply server/providers/FantLab.js:50]: Done

File: server/providers/FantLab.js (hunk 2 above)
Comment (Nab0y) [reply server/providers/FantLab.js]: Done

File: server/providers/FantLab.js (hunk 3 above)
Comment (Nab0y) [reply server/providers/FantLab.js:80]: Some books have covers, some don't. But anyway, I'm making a request for Editions. The priority is to get the cover of the audiobook, if there are no audiobooks in the publications, then there is a search for paper editions, immediately looking for isbn. If the book does not have any editions then the cover is taken from the description of the book, and if it does not exist, then nothing will be returned.

File: server/providers/FantLab.js (hunk 4 above)
Comment (Nab0y) [reply server/providers/FantLab.js]: Done

File: server/providers/FantLab.js (hunk 5 above)
Comment (Nab0y) [reply server/providers/FantLab.js:121]: image
root genre and sub genres has same node name

**Comment by Nab0y** [comment]
Code cleared of semicolons and var

File: server/providers/FantLab.js [hunk 6]
```diff
@@ -0,0 +1,172 @@
...
+    return Promise.all(items.map(async item => await this.getWork(item))).then(resArray => {
+      return resArray.filter(res => res != null)
+    })
+  }
+
+  async getWork(item) {
+    const { work_id, work_type_id } = item
+
+    if (this._filterWorkType.includes(work_type_id)) {
+      return { title: null }
```
Comment (advplyr) [inline server/providers/FantLab.js:50]:
```suggestion
      return null
```
On Line 42 null values are being filtered out.

File: server/providers/FantLab.js (hunk 3 above)
Comment (advplyr) [reply server/providers/FantLab.js:80]: Can you provide an example book or two I can look at?

Comment (Nab0y) [reply server/providers/FantLab.js:80]: 1. Title - Путь молодого бога
Author - Дмитрий Рус

api -  https://api.fantlab.ru/work/1283360/extended
site - https://fantlab.ru/work1283360
There is an image in the book data, there are also two editions, one audiobook and one paper book

2. Title - Дорога домой
Author - Дмитрий Рус

api - https://api.fantlab.ru/work/1333346/extended
site - https://fantlab.ru/work1333346
No image in the given book, three editions, two audiobooks and one paper book

3. Title - Принцесса где-то там
Author - Сергей Мусаниф

api - https://api.fantlab.ru/work/1636436/extended
site - https://fantlab.ru/work1636436
No images, no editions

4. Title - Цветы для Элджернона
Author - Дэниел Киз 
You may use English names Flowers for Algernon - Daniel Keyes result be the same :)

api - https://api.fantlab.ru/work/52178/extended
site - https://fantlab.ru/work52178
There is an image in book data, but also a bunch of editions, but no audiobooks. ISBN is being received

Comment (advplyr) [reply server/providers/FantLab.js]: Thanks for these test cases. Working well for these.

**Comment by advplyr** [comment]
Thanks!

**Comment by slyfoxgh** [comment]
Did the search stopped working. I can't match any Russian books anymore.

`audiobookshelf-audiobookshelf-1  | [2024-02-13 03:34:42.578] INFO: [Audnexus] Searching for author "https://api.audnex.us/authors?name=%D0%B1%D0%B5%D1%81%D1%82%D0%B5%D1%80"
audiobookshelf-audiobookshelf-1  | [2024-02-13 03:34:42.583] INFO: [Audnexus] Searching for author "https://api.audnex.us/authors?name=%D1%84%D0%B5%D0%BD%D0%BE%D0%BC%D0%B5%D0%BD%20%D0%B8%D1%81%D1%87%D0%B5%D0%B7%D0%BD%D0%BE%D0%B2%D0%B5%D0%BD%D0%B8%D1%8F"
audiobookshelf-audiobookshelf-1  | [2024-02-13 03:35:02.660] INFO: [Audnexus] Searching for author "https://api.audnex.us/authors?name=%D0%B1%D0%B5%D1%81%D1%82%D0%B5%D1%80"
audiobookshelf-audiobookshelf-1  | [2024-02-13 03:35:02.663] INFO: [Audnexus] Searching for author "https://api.audnex.us/authors?name=%D1%84%D0%B5%D0%BD%D0%BE%D0%BC%D0%B5%D0%BD%20%D0%B8%D1%81%D1%87%D0%B5%D0%B7%D0%BD%D0%BE%D0%B2%D0%B5%D0%BD%D0%B8%D1%8F"
`

**Pull request merged**